indexing-service/src/main/java/org/apache/druid/indexing/common/task/KillUnusedSegmentsTask.java

```diff
@@ -300,6 +356,81 @@ private NavigableMap<DateTime, List<TaskLock>> getNonRevokedTaskLockMap(TaskActi
     return taskLockMap;
   }
 
+  private List<DataSegment> findUnreferencedSegments(
+      List<DataSegment> unusedSegments,
+      Map<String, String> upgradedFromSegmentIds,
+      TaskActionClient taskActionClient
+  ) throws IOException
+  {
+    if (unusedSegments.isEmpty() || upgradedFromSegmentIds.isEmpty()) {
+      return unusedSegments;
+    }
+
+    final Set<String> upgradedSegmentIds = new HashSet<>();
+    for (DataSegment segment : unusedSegments) {
+      final String id = segment.getId().toString();
+      upgradedSegmentIds.add(id);
+      if (upgradedFromSegmentIds.containsKey(id)) {
+        upgradedSegmentIds.add(upgradedFromSegmentIds.get(id));
+      }
+    }
+
+    final Map<String, Set<String>> upgradedToSegmentIds;
+    try {
+      upgradedToSegmentIds = taskActionClient.submit(
+          new RetrieveUpgradedToSegmentIdsAction(getDataSource(), upgradedSegmentIds)
+      ).getUpgradedToSegmentIds();
+    }
+    catch (Exception e) {
+      LOG.warn(
+          e,
+          "Could not retrieve UpgradedToSegmentsResponse using task action[retrieveUpgradedToSegmentIds]."
+          + " Overlord may be on an older version."
+      );
+      return unusedSegments;
+    }
+
+    final Set<String> existingSegmentIds
+        = taskActionClient.submit(new RetrieveSegmentsByIdAction(getDataSource(), upgradedSegmentIds))
+                          .stream()
+                          .map(DataSegment::getId)
+                          .map(SegmentId::toString)
+                          .collect(Collectors.toSet());
+
+    List<DataSegment> unreferencedSegments = new ArrayList<>();
+    for (DataSegment segment : unusedSegments) {
```

Comment:
This loop and the whole method in general has become too complicated.

The method needs to do the following instead:

```java
private Set<DataSegment> findKillableSegments(...)
{

// Determine parentId for each unused segment
final Map<String, DataSegment> parentIdToUnusedSegment = new HashMap<>();
for (DataSegment segment : unusedSegments) {
    final String segmentId = segment.getId().toString();
    if (upgradedFromSegmentIds.containsKey(segmentId) {
        parentIdToUnusedSegment.put(upgradedFromSegmentIds.get(segmentId), segment);
    } else {
        parentIdToUnusedSegment.put(segmentId, segment);
    }
}

// Check if the parent or any of its children exist in metadata store
try {
    UpgradedToSegmentIdsResponse response = taskActionClient.submit(
          new RetrieveUpgradedToSegmentIdsAction(getDataSource(), parentIdToUnusedSegment.keySet())
    ).getUpgradedToSegmentIds();
    if (response != null && response.getUpgradedToSegmentIds() != null) {
        response.getUpgradedToSegmentIds().forEach((parent, children) -> {
             if (!CollectionUtils.isNullOrEmpty(children)) {
                 // Do not kill segment if its parent or any of its siblings still exist in metadata store
                 parentIdToUnusedSegment.remove(parent);
             }
        });
    }
}
catch (Exception e) {
      LOG.warn(
          e,
          "Could not retrieve UpgradedToSegmentsResponse using task action[retrieveUpgradedToSegmentIds]."
          + " Overlord may be on an older version."
);

// TODO: filter out based on used segment load specs

return parentIdToUnusedSegment.values();
}
```

Reply:
Thank you for attempting to refine this method.
I feel that the way it is written is intuitive to me at least and clearly lists out the cases in which a segment may or may not be referenced.
Also, the suggestion you have shared seems incorrect.
Please refer to KillUnusedSegmentsTaskTest with the changes in the latest commit.
I had made changes per your suggestion in a previous commit, found the logic to be inaccurate and reverted to my original approach. Perhaps the suggestion could also be tweaked to pass the test cases.

Reply:
> I feel that the way it is written is intuitive to me at least and clearly lists out the cases in which a segment may or may not be referenced.

Listing out the cases just makes the whole thing too verbose and confusing. There is only one case "we should not delete the segment files if any other segment is using the same files".

You can use your own method. I was just trying to demonstrate that the logic should be easy to follow with minimal, well-placed comments.

Reply:
Fixed. Thanks for the explanation